### ISSUE TITLE
Stack overflow parsing shift AST of lots of string concatenation

### ISSUE THREAD

**Issue by dominiccooney** [binast/binjs-ref#232]
Here's a 52 KB/700 line file file I encountered in a random sample of httparchive:
[f724.js.zip](https://github.com/binast/binjs-ref/files/2613698/f724.js.zip)

The nature of the file is:

````js
window["fpdefs"] = {
vertex_shader :
'attribute vec3 a_vertexPosition,a_vertexNormal;attribute vec2 a_vert' +
'exTexture;uniform mat4 u_mvMatrix,u_pMatrix,u_nMatrix;uniform vec3 u' +
...
texture :
'/9j/4AAQSkZJRgABAQEAAQABAAD/2wBDABALDA4MChAODQ4SERATGCgaGBYWGDEjJR0o' +
'OjM9PDkzODdASFxOQERXRTc4UG1RV19iZ2hnPk1xeXBkeFxlZ2P/wAALCAIAAgABAREA' +
...
````

This caused a stack overflow building dictionaries with d477a474500849c9c26abc51de27e7a399f5b77b:

````
Treating "/.../httparchive/chrome-Nov_1_2017/f724.js" ("")
Parsing.
thread 'large stack dedicated thread' panicked at 'Could not parse source: JsonError(ExceededDepthLimit)', libcore/result.rs:1009:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
thread 'main' panicked at 'Error in dedicated thread: Any', libcore/result.rs:1009:5
````

I guess when the Shift AST parser goes away, so will this JSON &rarr; AST step. Do you have a pointer to the Rust JavaScript parser you want to replace Shift with?

**Comment by Yoric** [comment]
As discussed somewhere else, one possible idea is to replace this with [ratel](https://github.com/ratel-rust/ratel-core).

**Comment by Yoric** [comment]
See #229 (inactive atm).

**Comment by Yoric** [comment]
I can confirm that the problem still appears as of tag entropy-0.2.0.

**Comment by Yoric** [comment]
Cc @RReverser that's a demonstration of the kind of issues we encounter with JSON parsing in real examples already.

**Comment by RReverser** [comment]
Understood. @dominiccooney Note however that replacing with another parser won't alleviate the problem, it can just increase the limit at which a fatal stack overflow occurs.

**Comment by dominiccooney** [comment]
@RReverser Yeah I understand that. FWIW I don't think Ratel tip has been able to parse fb_bench for a while. (Maybe I'm holding it wrong.)
